### PR TITLE
[SelectionDAG] Add SDNode::user_begin() and use it in some places

### PR DESCRIPTION
**Pull request by topperc**
Most of these are just places that want the first user and aren't
iterating over the whole list.

While there I changed some use_size() == 1 to hasOneUse() which
is more efficient.

This is part of an effort to rename use_iterator to user_iterator
and provide a use_iterator that dereferences to SDUse. This patch
helps reduce the diff on later patches.

This is stacked on #120499

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -10906,15 +10906,15 @@ SDValue DAGCombiner::visitSRL(SDNode *N) {
...
   // processed in topological order.
   if (N->hasOneUse()) {
-    SDNode *Use = *N->use_begin();
+    SDNode *User = *N->user_begin();
 
     // Look pass the truncate.
-    if (Use->getOpcode() == ISD::TRUNCATE && Use->hasOneUse())
-      Use = *Use->use_begin();
+    if (User->getOpcode() == ISD::TRUNCATE && User->hasOneUse())
+      User = *User->user_begin();
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10913]: Should have a getOneUser helper

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10913]: Is there an equivalent in IR?

**Review by s-barannikov — APPROVED** [review]
LGTM

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10913]: The IR has hasOneUser, which is strictly less useful than returning the user.

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10913]: hasOneUser in IR returns whether or not the User of all Uses is unique. Should your proposed getOneUser() return the unique User or all Uses? Or should it only return the User if there is exactly one Use? Most of these places probably expect the latter.

Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10913]: Correct, the point is to handle instructions where the same value is used multiple times. It's a single user, which is usually what you want. It returns the single unique instruction which may use the value multiple times as an operand

**Pull request merged**